src/Http/Controllers/WebhookController.php

```diff
@@ -163,11 +164,14 @@ protected function handleCustomerSubscriptionUpdated(array $payload)
                 }
             }
 
+            // Renewal date...
+            $subscription->renews_at = Carbon::createFromTimestamp($data['current_period_end']);
```

Comment:
Is this webhook endpoint invoked on every billing cycle's successful payment?

Reply:
The `customer.subscription.updated` webhook is triggered when the billing period ends, usually an hour before the renewal payment is attempted.

It is triggered again if the payment fails to set the subscription status to "past_due".

So it'll keep Cashier's `renews_at` column up to date with Stripe subscriptions' billing periods.

However, the one hour delay that might cause confusion. A user could see a future renewal date thinking the payment went through when it could actually fail a few minutes later.
A workaround would be to update `renews_at` inside another webhook instead (probably `invoice.payment_succeeded`).

<details>
<summary>These are the webhook events I recorded</summary>

Successful subscription renewal:
- `invoice.created`
- `customer.subscription.updated` (new `current_period_start|end` are set)
- [~ 1 hour delay]
- `payment_intent.created`
- `payment_intent.succeeded`
- `customer.updated`
- `invoice.updated` (status is "paid")
- `invoice.paid`
- `invoice.payment_succeeded`
- `invoice.finalized`

Failed subscription renewal:
- `invoice.created`
- `customer.subscription.updated` (new `current_period_start|end` are set)
- [~ 1 hour delay]
- `payment_intent.created`
- `payment_intent.payment_failed`
- `customer.updated`
- `invoice.updated` (status is "open")
- `customer.subscription.updated` (set subscription status to "past_due")
- `invoice.payment_failed`
- `invoice.updated`
- `invoice.finalized`

Successful subscription swap:
- `payment_intent.created`
- `payment_intent.succeeded`
- `invoiceitem.created` (proration adjustment)
- `invoice.created`
- `invoice.paid`
- `invoice.payment_succeeded`
- `invoice.finalized`
- `customer.subscription.updated`

Successful subscription creation:
- `payment_intent.created`
- `invoice.created`
- `invoice.finalized` (status is "open")
- `customer.subscription.created`
- `payment_method.attached`
- `payment_intent.succeeded`
- `invoice.updated` (status is "paid")
- `invoice.paid`
- `invoice.payment_succeeded`
</details>

Reply:
@clementmas I don't understand sorry. From your examples above, `customer.subscription.updated` is always triggered on time with the new `current_period_start|end` set? I don't see why we should listen to a different webhook? Subscription created is also covered with its own event.

Reply:
The webhooks for new and swapped subscriptions are triggered on time but there's a delay for renewed subscriptions.

The `customer.subscription.updated` webhook is triggered an hour before the payment to renew the subscription is attempted.

<details>
<summary>See webhook events with timestamps</summary>

Apr 1, 2024, 1:15:06 PM: `invoice.upcoming`
Apr 7, 2024, 1:14:27 PM: `invoice.created`
Apr 7, 2024, 1:14:27 PM: `customer.subscription.updated`
Apr 7, 2024, 2:15:18 PM: `payment_intent.created`
Apr 7, 2024, 1:15:19 PM: `payment_intent.succeeded`
...
</details>

So during that hour, the renewal date is technically wrong. It should display "today"'s date instead of next month/year.

But if that's not a concern, then yes, it's much simpler to keep relying on the `customer.subscription.created` and `customer.subscription.updated` webhooks.

Reply:
I wouldn't worry about that. I'd always rely on the `customer.subscription.updated` event.

Reply:
I think it's important that we mimic exactly what Stripe has as a value on the subscription. If the renewal date is adjusted an hour before the payment is even attempted then we should take that over no matter what happens after it. If the payment fails and anything changes on the Subscription entity we'll get a new updated event with the new values.

Reply:
@driesvints I think we should focus more on what the user would expect to see. Do we not have way to show a truly accurate renewal date?

Reply:
I think we have to change a few things here as period_ends_at from Stripe is not same as renews_at.
For example in the case of a cancellation the updated webhook is called and we should set renews_at to null instead of the end of the period.

We could do this this way :
```php
       // Period ends at
        $currentPeriodEndsAt = Carbon::createFromTimestamp($data['current_period_end']);

        // Cancellation date...
        if ($data['cancel_at_period_end'] ?? false) {
            $subscription->ends_at = $subscription->onTrial()
                ? $subscription->trial_ends_at
                : $subscription->renews_at;
            $subscription->renews_at = null;
        } elseif (isset($data['cancel_at']) || isset($data['canceled_at'])) {
            $subscription->ends_at = Carbon::createFromTimestamp($data['cancel_at'] ?? $data['canceled_at']);
            $subscription->renews_at = null;
        } else {
            $subscription->ends_at = null;
            $subscription->renews_at = $currentPeriodEndsAt;
        }
```

Reply:
I think we have to change a few things here as period_ends_at from Stripe is not same as renews_at.
For example in the case of a cancellation the updated webhook is called and we should set renews_at to null instead of the end of the period.

We could do this this way :
```php
        // Period ends at
        $currentPeriodEndsAt = Carbon::createFromTimestamp($data['current_period_end']);

        // Cancellation date...
        if ($data['cancel_at_period_end'] ?? false) {
            $subscription->renews_at = null;
            $subscription->ends_at = $subscription->onTrial()
                ? $subscription->trial_ends_at
                : $;currentPeriodEndsAt;
        } elseif (isset($data['cancel_at']) || isset($data['canceled_at'])) {
            $subscription->renews_at = null;
            $subscription->ends_at = Carbon::createFromTimestamp($data['cancel_at'] ?? $data['canceled_at']);
        } else {
            // Renewal date...
            $subscription->renews_at = $currentPeriodEndsAt;
            $subscription->ends_at = null;
        }
```

Reply:
@ybert I now wonder if `billing_cycle_anchor` is more what we need? https://docs.stripe.com/billing/subscriptions/billing-cycle

Although I don't know what happens for metered billing.

Reply:
@driesvints I just tested it, and in my use cases, current_period_end and billing_cycle_anchor are the same after cancellation.

Reply:
@taylorotwell but what Stripe shows IS the accurate date at the time the event is shown. We have no way of knowing when to change the value if we delay it until the payment has been done as that's a complex process with lots of unknown and side effects that could happen. The fact that the renewal date changes when a payment fails is such a side effect. We should simple focus on reflecting what Stripe tells us. `current_period_end` will always be the correct indication when a subscription renews.

@ybert yeah I agree with the above code, feel free to adopt that in your PR.

Reply:
Thanks @clementmas. I just feel there's gonna be quite a bit of side effects from relying on the payment succeeded webhook. When you're ready feel free to try this but I'd suggest to create a new draft PR so we can inspect the differences.